tests/sentry/web/frontend/test_oauth_token.py

```diff
@@ -115,6 +115,89 @@ def test_invalid_code(self):
         assert resp.status_code == 400
         assert json.loads(resp.content) == {"error": "invalid_grant"}
 
+    def test_expired_grant(self):
+        self.login_as(self.user)
+        expired_grant = ApiGrant.objects.create(
+            user=self.user,
+            application=self.application,
+            redirect_uri="https://example.com",
+            expires_at="2022-01-01 11:11",
+        )
+        resp = self.client.post(
+            self.path,
+            {
+                "grant_type": "authorization_code",
+                "redirect_uri": self.application.get_default_redirect_uri(),
+                "code": expired_grant.code,
+                "client_id": self.application.client_id,
+            },
+        )
+        assert resp.status_code == 400
+        assert json.loads(resp.content) == {"error": "invalid_grant"}
+
+    def test_invalid_redirect_uri(self):
+        self.login_as(self.user)
+
+        resp = self.client.post(
+            self.path,
+            {
+                "grant_type": "authorization_code",
+                "code": self.grant.code,
+                "client_id": self.application.client_id,
+                "redirect_uri": "cheese.org",
+            },
+        )
+        assert resp.status_code == 400
+        assert json.loads(resp.content) == {"error": "invalid_grant"}
+
+    def test_no_open_id_token(self):
+        """
+        Checks that the OIDC token is not returned unless the right scope is approved.
+        """
+        self.login_as(self.user)
+
+        resp = self.client.post(
+            self.path,
+            {
+                "grant_type": "authorization_code",
+                "code": self.grant.code,
+                "client_id": self.application.client_id,
+            },
+        )
+
+        assert resp.status_code == 200
+        data = json.loads(resp.content)
+        assert "id_token" not in data
+
+    def test_valid_no_redirect_uri(self):
+        """
+        Checks that we get the correct redirect URI if we don't pass one in
+        """
+        self.login_as(self.user)
+
+        resp = self.client.post(
+            self.path,
+            {
+                "grant_type": "authorization_code",
+                "code": self.grant.code,
+                "client_id": self.application.client_id,
+            },
+        )
+
+        assert resp.status_code == 200
+        data = json.loads(resp.content)
+
+        token = ApiToken.objects.get(token=data["access_token"])
+        assert token.application == self.application
+        assert token.user == self.grant.user
+        assert token.get_scopes() == self.grant.get_scopes()
+
+        assert data["access_token"] == token.token
+        assert data["refresh_token"] == token.refresh_token
+        assert isinstance(data["expires_in"], int)
+        assert data["token_type"] == "bearer"
+        assert data["user"]["id"] == str(token.user_id)
```

Comment:
Should this test do assertions on the redirect uri? The test name implies that would  happen.

Reply:
There's no redirect URI or anything returned here - basically we're checking that whatever redirect URI was passed to `/authorize` to generate the authorization code is also passed to `/token`. So if we don't pass a redirect URI to `/authorize` we use the default, and therefore we don't need to pass one to `/token` since it will check against the default URI. This test just makes sure it doesn't error out in that case.